### PR TITLE
feat(app): forward gap-fill backfill pass for missed sessions (#627)

### LINKED ISSUE #627: Bidirectional backfill: forward gap-fill pass for missed sessions
## Parent

Implements the forward gap-fill for #626. Design locked there.

## What to build

Extend the backfill job with a **forward pass** so a gap between the newest stored point and now — a session the app missed while down — is recovered from `history()`. Today backfill only walks backward (oldest → first BUY) and marks the series `complete`; nothing fills the recent end, so any downtime during `REGULAR` is a permanent hole.

Each backfill cycle, per (symbol, account):

- Read the **newest** stored timestamp — a new `get_newest_timestamp`, mirror of `get_oldest_timestamp`, scoped by `share_symbol` + `account`.
- Compute the forward window `[newest, now]`; if it spans real trading time, fetch it (`1h`, `1d` beyond 730d — same granularity as the backward pass, chunked the same way) and write the returned rows into `portfolio_metrics` with the **same series identity/tags** as live and backward-filled points.
- **Self-classify by delegating to yfinance:** write whatever rows `history()` returns. An empty window (weekend/holiday, or already covered) writes nothing and is a no-op. No calendar logic, no gap classifier.
- Keep the window-sizing decision in a **pure function** (in `scheduling.py`, mirroring the existing pure-module pattern) so it is unit-testable against an injected `now` with no InfluxDB/yfinance.
- The forward pass must **no-op during live trading** — newest ≈ now → sub-day window → skip (reuse the existing `< 1 day` guard) — so it never races or duplicates the live `REGULAR` writer, which stays the sole writer of the present.
- The backward-completion watermark (`_backfill_complete`) must **not** suppress the forward pass; the two directions are independent.

## Acceptance criteria

- [ ] `get_newest_timestamp(symbol, account)` returns the latest stored point's time (or `None`), scoped by symbol + account like `get_oldest_timestamp`.
- [ ] After a simulated gap (stop the app across a trading session, restart), the missed session's `1h` points appear in `portfolio_metrics` within a cycle or two.
- [ ] A weekend/holiday gap stays empty (yfinance returns no rows → nothing written).
- [ ] During live `REGULAR` trading the forward pass writes nothing and produces no duplicate points at the seam.
- [ ] The window-sizing logic is a pure function with unit tests (injected `now`, no network/DB): fresh gap, no gap, weekend-empty, multi-chunk long gap.
- [ ] Recovered points carry the same tags/series identity as live points, so the perf series `holdings_value` picks them up on the next recompute.
- [ ] `flake8` clean; existing backward-pass behavior and `_backfill_complete` handling unchanged.

## Blocked by

None - can start immediately.

### PR DESCRIPTION
**Pull request by pbrissaud**
## Résumé

Implémente le **forward gap-fill** du backfill bidirectionnel (issue #627, slice 1 de #626). Le backfill ne parcourait que le passé (oldest → premier BUY) puis marquait la série `complete` ; toute indisponibilité pendant une session `REGULAR` laissait donc un trou permanent. Cette PR ajoute une **passe forward indépendante** sur `[newest, now]` qui **délègue le calendrier à yfinance** (aucune logique de jours fériés/week-ends).

Contexte détaillé : issue #626 (décisions verrouillées).

## Changements

- **`influxdb_writer.py`** — `get_newest_timestamp(symbol, account)`, miroir de `get_oldest_timestamp` (`ORDER BY time DESC`), même scoping `COALESCE(account, 'default')`.
- **`scheduling.py`** — `forward_backfill_window(newest, now, chunk_days)`, fonction **pure** (`now` injecté, sans InfluxDB/yfinance). Réutilise le garde-fou `< 1 jour` → **no-op pendant le trading live** (newest ≈ now, aucun doublon au raccord) ; chunké comme la passe backward.
- **`main.py`** — `backfill()` refactorisé en `_backfill_share` → `_backfill_backward` / `_backfill_forward`, avec les helpers partagés `_ensure_share_info`, `_enrich_and_write`, `_fetch_and_store`. Le watermark `_backfill_complete` ne gate **que** la passe backward ; les deux directions sont indépendantes. Les points récupérés portent les mêmes tags/identité de série que les points live, donc `holdings_value` (perf) les reprend au prochain recompute.

## Tests

- `forward_backfill_window` : fresh gap, no gap, sous-jour/live, multi-chunk, newest None, tz naïf, clock-skew.
- `get_newest_timestamp` : retour, empty, exception, `ORDER BY DESC`, scoping account, échappement quotes.
- Intégration backfill : comblement d'un gap récent, fenêtre vide = no-op, série vide = no-op, deux directions dans un même cycle.
- **439 tests passent**, **flake8 propre**. Comportement backward-pass et `_backfill_complete` inchangés.

Closes #627

🤖 Generated with [Claude Code](https://claude.com/claude-code)

https://claude.ai/code/session_01Aq4dzKFBnKwmEfr4UouXhr

## Summary by CodeRabbit

* **New Features**
  * Added forward gap-filling to recover recent missing market data using a “newest stored timestamp” anchor.
  * Enhanced backfill to run independent backward (older gaps) and forward (recent gaps) passes per share.
* **Bug Fixes**
  * Prevent forward recovery during live trading or sub-day/no-gap windows, and handle clock-skew/timezone-naive anchors safely.
  * Improved enrichment/write flow so failures for one share don’t interrupt the full backfill.
* **Tests**
  * Expanded coverage for newest-timestamp lookup, forward window sizing, forward/backward backfill interactions, and error/edge cases.

### CODE REVIEW

**Comment by pbrissaud** [comment]
Merci @coderabbitai. Dispositions (commit fd00c4a) :

**✅ Corrigé — `share_price IS NOT NULL` sur `get_newest_timestamp`.** Juste : `write_metrics` peut persister un point sans prix (`influxdb_writer.py:156`), et la passe forward comble des trous de *prix* que lit `holdings_value` (`get_price_series` filtre pareil). L'ancre porte désormais sur le point le plus récent *porteur de prix*.

**✅ Corrigé — docstring de `backfill()`** mise à jour pour décrire les deux passes (backward + forward indépendante).

**❌ Rejeté — scoping par compte sur `get_newest_timestamp`.** Le critère d'acceptation de #627 impose explicitement « *scoped by `share_symbol` + `account` like `get_oldest_timestamp`* ». La règle « market price by symbol only » vise `get_price_series` (un prix de marché n'appartient à aucun compte), pas une **ancre de couverture** de backfill qui est par-série : chaque compte doit voir son propre trou. Symbol-only sous-remplirait un compte plus récent partageant un symbole.

**❌ Rejeté — `time.sleep` non mocké dans `test_backfill_forward_fills_recent_gap`.** Sans objet : la fixture `autouse` `_no_sleep` (`test_metrics.py:91`) neutralise déjà `time.sleep` pour tout le module ; les `mocker.patch("main.time.sleep")` des tests voisins sont en fait redondants.

File: app/src/influxdb_writer.py [hunk 1]
```diff
@@ -326,6 +326,60 @@ def get_oldest_timestamp(
...
+        """
+        if self._client is None:
+            self.connect()
+
+        # Escape single quotes to keep share_symbol a safe SQL string literal
+        safe_symbol = share_symbol.replace("'", "''")
+        where = f"share_symbol = '{safe_symbol}'"
+        if account is not None:
+            safe_account = account.replace("'", "''")
+            where += f" AND COALESCE(account, 'default') = '{safe_account}'"
```
[the remark this answers is not in the document]
Comment (pbrissaud) [reply app/src/influxdb_writer.py:367]: Je maintiens le scoping par compte — après vérification, la guideline citée ne s'applique pas à cette fonction, et le symbol-only introduirait le bug qu'elle prétend éviter.

**La guideline est propre à `get_price_series`.** Dans `CLAUDE.md:365-366` : « *Price history for holdings_value is read via `InfluxDBWriter.get_price_series(symbol)` — queried by `share_symbol` only, never by `account`* ». Elle vise la lecture des **prix** consommés par `holdings_value` (où filtrer par compte tronquerait l'historique, les points pré-tag ayant `account=NULL`). `get_newest_timestamp` ne lit pas un prix pour le consommer : c'est une **ancre de couverture** qui répond à « où se termine la série de *ce* couple (symbol, account) ? ».

**Le sibling imposé par la spec est déjà account-scoped, volontairement.** `get_oldest_timestamp` (`influxdb_writer.py:287`) : « *scope the lookup to this account so backfill gaps are detected per account* ». Le critère d'acceptation de #627 demande explicitement « *scoped by `share_symbol` + `account` like `get_oldest_timestamp`* ». Une ancre de couverture est par-série, pas par-prix-de-marché.

**Symbol-only causerait la divergence qu'il prétend corriger.** Contre-exemple : comptes A et B détiennent AAPL, le scrape de B échoue transitoirement lundi mais celui de A passe → `newest_A = lundi 10:00`, `newest_B = vendredi`.
- Account-scoped : B comble `[vendredi, now]` → rattrape son lundi. ✅
- Symbol-only : B voit `newest = lundi 10:00` (celui de A) → le trou de B `[vendredi, lundi 10:00]` n'est jamais comblé → **B diverge**. ❌

De même, un nouveau compte sans point renvoie `None` (account-scoped) → forward no-op, la passe backward sème la série (correct). En symbol-only il hériterait du `newest` d'un autre compte et écrirait des points enrichis d'un état à zéro pour une période non détenue.

Le scoping par compte est donc strictement plus correct pour une ancre de couverture. Je laisse le code inchangé.
